File: backend/app/services/certificate_service.py
import os
import uuid
import time
from typing import Dict, List, Optional
from pydantic import BaseModel
from datetime import datetime, timedelta
import random

class Certificate(BaseModel):
    id: str
    user_id: str
    kurban_id: str
    certificate_type: str  # "kurban", "bagis", "katilim"
    title: str
    description: str
    issued_date: str
    expiry_date: Optional[str] = None
    status: str  # "active", "expired", "revoked"
    verification_code: str
    qr_code_url: Optional[str] = None
    pdf_url: Optional[str] = None
    metadata: Dict = {}

class CertificateService:
    def __init__(self):
        self.certificates: Dict[str, Certificate] = {}
        self._initialize_mock_data()
    
    def _initialize_mock_data(self):
        """Mock sertifika verilerini başlat"""
        mock_certificates = [
            {
                "id": "cert_001",
                "user_id": "user_123",
                "kurban_id": "kurban_001",
                "certificate_type": "kurban",
                "title": "Kurban Katılım Sertifikası",
                "description": "2024 Kurban Bayramı kurban kesimi katılım sertifikası",
                "issued_date": "2024-06-15",
                "expiry_date": "2025-06-15",
                "status": "active",
                "verification_code": "KC2024001",
                "qr_code_url": "https://api.qrserver.com/v1/create-qr-code/?size=200x200&data=KC2024001",
                "pdf_url": "/api/certificates/cert_001/download",
                "metadata": {
                    "animal_type": "koyun",
                    "weight": "45kg",
                    "location": "Ankara",
                    "imam": "Mehmet Yılmaz"
                }
            },
            {
                "id": "cert_002",
                "user_id": "user_456",
                "kurban_id": "kurban_002",
                "certificate_type": "bagis",
                "title": "Bağış Katılım Sertifikası",
                "description": "Kurban bağışı katılım sertifikası",
                "issued_date": "2024-06-16",
                "expiry_date": None,
                "status": "active",
                "verification_code": "KC2024002",
                "qr_code_url": "https://api.qrserver.com/v1/create-qr-code/?size=200x200&data=KC2024002",
                "pdf_url": "/api/certificates/cert_002/download",
                "metadata": {
                    "amount": 500.0,
                    "currency": "TL",
                    "payment_method": "kredi_karti"
                }
            },
            {
                "id": "cert_003",
                "user_id": "user_789",
                "kurban_id": "kurban_003",
                "certificate_type": "katilim",
                "title": "Etkinlik Katılım Sertifikası",
                "description": "Kurban organizasyonu katılım sertifikası",
                "issued_date": "2024-06-17",
                "expiry_date": "2025-06-17",
                "status": "active",
                "verification_code": "KC2024003",
                "qr_code_url": "https://api.qrserver.com/v1/create-qr-code/?size=200x200&data=KC2024003",
                "pdf_url": "/api/certificates/cert_003/download",
                "metadata": {
                    "event_name": "Kurban Organizasyonu 2024",
                    "location": "İstanbul",
                    "duration": "3 gün"
                }
            }
        ]
        
        for cert_data in mock_certificates:
            certificate = Certificate(**cert_data)
            self.certificates[certificate.id] = certificate
    
    async def create_certificate(self, 
                               user_id: str,
                               kurban_id: str,
                               certificate_type: str,
                               title: str,
                               description: str,
                               metadata: Dict = {}) -> Certificate:
        """Yeni sertifika oluştur"""
        cert_id = f"cert_{uuid.uuid4().hex[:8]}"
        verification_code = f"KC{int(time.time())}{random.randint(100, 999)}"
        
        certificate = Certificate(
            id=cert_id,
            user_id=user_id,
            kurban_id=kurban_id,
            certificate_type=certificate_type,
            title=title,
            description=description,
            issued_date=datetime.now().strftime("%Y-%m-%d"),
            expiry_date=(datetime.now() + timedelta(days=365)).strftime("%Y-%m-%d") if certificate_type != "bagis" else None,
            status="active",
            verification_code=verification_code,
            qr_code_url=f"https://api.qrserver.com/v1/create-qr-code/?size=200x200&data={verification_code}",
            pdf_url=f"/api/certificates/{cert_id}/download",
            metadata=metadata
        )
        
        self.certificates[cert_id] = certificate
        return certificate
    
    async def get_certificate(self, cert_id: str) -> Optional[Certificate]:
        """Sertifika getir"""
        return self.certificates.get(cert_id)
    
    async def get_user_certificates(self, user_id: str) -> List[Certificate]:
        """Kullanıcının sertifikalarını getir"""
        return [cert for cert in self.certificates.values() if cert.user_id == user_id]
    
    async def verify_certificate(self, verification_code: str) -> Optional[Certificate]:
        """Sertifika doğrula"""
        for cert in self.certificates.values():
            if cert.verification_code == verification_code:
                return cert
        return None
    
    async def revoke_certificate(self, cert_id: str) -> bool:
        """Sertifikayı iptal et"""
        if cert_id in self.certificates:
            self.certificates[cert_id].status = "revoked"
            return True
        return False
    
    async def get_all_certificates(self, 
                                 skip: int = 0, 
                                 limit: int = 100,
                                 status: Optional[str] = None,
                                 certificate_type: Optional[str] = None) -> List[Certificate]:
        """Tüm sertifikaları getir (filtreleme ile)"""
        certificates = list(self.certificates.values())
        
        if status:
            certificates = [cert for cert in certificates if cert.status == status]
        
        if certificate_type:
            certificates = [cert for cert in certificates if cert.certificate_type == certificate_type]
        
        # Sıralama (en yeni önce)
        certificates.sort(key=lambda x: x.issued_date, reverse=True)
        
        return certificates[skip:skip + limit]
    
    async def get_certificate_stats(self) -> Dict:
        """Sertifika istatistikleri"""
        total = len(self.certificates)
        active = len([cert for cert in self.certificates.values() if cert.status == "active"])
        expired = len([cert for cert in self.certificates.values() if cert.status == "expired"])
        revoked = len([cert for cert in self.certificates.values() if cert.status == "revoked"])
        
        by_type = {}
        for cert in self.certificates.values():
            cert_type = cert.certificate_type
            by_type[cert_type] = by_type.get(cert_type, 0) + 1
        
        return {
            "total": total,
            "active": active,
            "expired": expired,
            "revoked": revoked,
            "by_type": by_type
        }
    
    async def generate_certificate_pdf(self, cert_id: str) -> Dict:
        """Sertifika PDF'i oluştur (mock)"""
        certificate = await self.get_certificate(cert_id)
        if not certificate:
            return {"success": False, "error": "Sertifika bulunamadı"}
        
        # Mock PDF oluşturma
        pdf_data = {
            "certificate_id": cert_id,
            "title": certificate.title,
            "verification_code": certificate.verification_code,
            "issued_date": certificate.issued_date,
            "qr_code": certificate.qr_code_url,
            "metadata": certificate.metadata
        }
        
        return {
            "success": True,
            "pdf_url": certificate.pdf_url,
            "data": pdf_data,
            "message": "PDF başarıyla oluşturuldu"
        }

# Global instance
certificate_service = CertificateService()

File: kurban-cebimde/services/CertificateService.ts
import AsyncStorage from '@react-native-async-storage/async-storage';

export interface Certificate {
  id: string;
  user_id: string;
  kurban_id: string;
  certificate_type: 'kurban' | 'bagis' | 'katilim';
  title: string;
  description: string;
  issued_date: string;
  expiry_date?: string;
  status: 'active' | 'expired' | 'revoked';
  verification_code: string;
  qr_code_url?: string;
  pdf_url?: string;
  metadata: Record<string, any>;
}

export interface CertificateStats {
  total: number;
  active: number;
  expired: number;
  revoked: number;
  by_type: Record<string, number>;
}

export interface CreateCertificateRequest {
  user_id: string;
  kurban_id: string;
  certificate_type: 'kurban' | 'bagis' | 'katilim';
  title: string;
  description: string;
  metadata?: Record<string, any>;
}

export class CertificateService {
  private static instance: CertificateService;
  private baseUrl: string;

  private constructor() {
    this.baseUrl = `${process.env.EXPO_PUBLIC_API_BASE}/api/certificates/v1`;
  }

  public static getInstance(): CertificateService {
    if (!CertificateService.instance) {
      CertificateService.instance = new CertificateService();
    }
    return CertificateService.instance;
  }

  /**
   * Yeni sertifika oluştur
   */
  public async createCertificate(request: CreateCertificateRequest): Promise<Certificate | null> {
    try {
      const response = await fetch(`${this.baseUrl}/create`, {
        method: 'POST',
        headers: {
          'Content-Type': 'application/json',
        },
        body: JSON.stringify(request),
      });

      if (response.ok) {
        const result = await response.json();
        return result.data;
      } else {
        console.error('Sertifika oluşturulamadı:', response.status);
        return null;
      }
    } catch (error) {
      console.error('Sertifika oluşturulurken hata:', error);
      return null;
    }
  }

  /**
   * Sertifika detaylarını getir
   */
  public async getCertificate(certId: string): Promise<Certificate | null> {
    try {
      const response = await fetch(`${this.baseUrl}/${certId}`);

      if (response.ok) {
        const result = await response.json();
        return result.data;
      } else if (response.status === 404) {
        console.log('Sertifika bulunamadı');
        return null;
      } else {
        console.error('Sertifika getirilemedi:', response.status);
        return null;
      }
    } catch (error) {
      console.error('Sertifika getirilirken hata:', error);
      return null;
    }
  }

  /**
   * Kullanıcının sertifikalarını getir
   */
  public async getUserCertificates(userId: string): Promise<Certificate[]> {
    try {
      const response = await fetch(`${this.baseUrl}/user/${userId}`);

      if (response.ok) {
        const result = await response.json();
        return result.data;
      } else {
        console.error('Kullanıcı sertifikaları getirilemedi:', response.status);
        return [];
      }
    } catch (error) {
      console.error('Kullanıcı sertifikaları getirilirken hata:', error);
      return [];
    }
  }

  /**
   * Sertifika doğrula
   */
  public async verifyCertificate(verificationCode: string): Promise<Certificate | null> {
    try {
      const response = await fetch(`${this.baseUrl}/verify/${verificationCode}`);

      if (response.ok) {
        const result = await response.json();
        return result.data;
      } else if (response.status === 404) {
        console.log('Geçersiz doğrulama kodu');
        return null;
      } else {
        console.error('Sertifika doğrulanamadı:', response.status);
        return null;
      }
    } catch (error) {
      console.error('Sertifika doğrulanırken hata:', error);
      return null;
    }
  }

  /**
   * Sertifikayı iptal et
   */
  public async revokeCertificate(certId: string): Promise<boolean> {
    try {
      const response = await fetch(`${this.baseUrl}/${certId}/revoke`, {
        method: 'DELETE',
      });

      if (response.ok) {
        console.log('Sertifika iptal edildi');
        return true;
      } else if (response.status === 404) {
        console.log('Sertifika bulunamadı');
        return false;
      } else {
        console.error('Sertifika iptal edilemedi:', response.status);
        return false;
      }
    } catch (error) {
      console.error('Sertifika iptal edilirken hata:', error);
      return false;
    }
  }

  /**
   * Tüm sertifikaları getir (filtreleme ile)
   */
  public async getAllCertificates(
    skip: number = 0,
    limit: number = 100,
    status?: string,
    certificateType?: string
  ): Promise<Certificate[]> {
    try {
      const params = new URLSearchParams({
        skip: skip.toString(),
        limit: limit.toString(),
      });

      if (status) params.append('status', status);
      if (certificateType) params.append('certificate_type', certificateType);

      const response = await fetch(`${this.baseUrl}/?${params}`);

      if (response.ok) {
        const result = await response.json();
        return result.data;
      } else {
        console.error('Sertifikalar getirilemedi:', response.status);
        return [];
      }
    } catch (error) {
      console.error('Sertifikalar getirilirken hata:', error);
      return [];
    }
  }

  /**
   * Sertifika istatistikleri
   */
  public async getCertificateStats(): Promise<CertificateStats | null> {
    try {
      const response = await fetch(`${this.baseUrl}/stats/overview`);

      if (response.ok) {
        const result = await response.json();
        return result.data;
      } else {
        console.error('İstatistikler getirilemedi:', response.status);
        return null;
      }
    } catch (error) {
      console.error('İstatistikler getirilirken hata:', error);
      return null;
    }
  }

  /**
   * Sertifika PDF'ini indir
   */
  public async downloadCertificatePdf(certId: string): Promise<string | null> {
    try {
      const response = await fetch(`${this.baseUrl}/${certId}/download`);

      if (response.ok) {
        const result = await response.json();
        return result.download_url;
      } else {
        console.error('PDF indirilemedi:', response.status);
        return null;
      }
    } catch (error) {
      console.error('PDF indirilirken hata:', error);
      return null;
    }
  }

  /**
   * Örnek sertifikaları getir (test için)
   */
  public async getSampleCertificates(): Promise<Certificate[]> {
    try {
      const response = await fetch(`${this.baseUrl}/mock/sample`);

      if (response.ok) {
        const result = await response.json();
        return result.data;
      } else {
        console.error('Örnek sertifikalar getirilemedi:', response.status);
        return [];
      }
    } catch (error) {
      console.error('Örnek sertifikalar getirilirken hata:', error);
      return [];
    }
  }

  /**
   * Mock sertifika oluştur (test için)
   */
  public async generateMockCertificate(): Promise<Certificate | null> {
    try {
      const response = await fetch(`${this.baseUrl}/mock/generate`, {
        method: 'POST',
      });

      if (response.ok) {
        const result = await response.json();
        return result.data;
      } else {
        console.error('Mock sertifika oluşturulamadı:', response.status);
        return null;
      }
    } catch (error) {
      console.error('Mock sertifika oluşturulurken hata:', error);
      return null;
    }
  }

  /**
   * Sertifikaları local storage'a kaydet
   */
  public async saveCertificatesToStorage(certificates: Certificate[]): Promise<void> {
    try {
      await AsyncStorage.setItem('user_certificates', JSON.stringify(certificates));
    } catch (error) {
      console.error('Sertifikalar kaydedilemedi:', error);
    }
  }

  /**
   * Local storage'dan sertifikaları al
   */
  public async getCertificatesFromStorage(): Promise<Certificate[]> {
    try {
      const stored = await AsyncStorage.getItem('user_certificates');
      return stored ? JSON.parse(stored) : [];
    } catch (error) {
      console.error('Sertifikalar alınamadı:', error);
      return [];
    }
  }

  /**
   * Sertifika tipine göre ikon getir
   */
  public getCertificateIcon(type: string): string {
    switch (type) {
      case 'kurban':
        return '🐑';
      case 'bagis':
        return '💰';
      case 'katilim':
        return '📜';
      default:
        return '📄';
    }
  }

  /**
   * Sertifika durumuna göre renk getir
   */
  public getCertificateStatusColor(status: string): string {
    switch (status) {
      case 'active':
        return '#4CAF50'; // Yeşil
      case 'expired':
        return '#FF9800'; // Turuncu
      case 'revoked':
        return '#F44336'; // Kırmızı
      default:
        return '#9E9E9E'; // Gri
    }
  }

  /**
   * Sertifika durumuna göre metin getir
   */
  public getCertificateStatusText(status: string): string {
    switch (status) {
      case 'active':
        return 'Aktif';
      case 'expired':
        return 'Süresi Dolmuş';
      case 'revoked':
        return 'İptal Edilmiş';
      default:
        return 'Bilinmiyor';
    }
  }
}

// Singleton instance export
export const certificateService = CertificateService.getInstance();
